Feature/add behaviors extension

bijington opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Markup Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Markup Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI Markup core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui.markup/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

 ### Description of Change ###

 <!-- Describe your changes here. This only needs to be brief as the linked issues below will already cover the detailed changes. -->

Adds a new `Behaviors` extension to `VisualElement`s to allow developers to fluently set `Behavior`s.

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #29

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui.Markup/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/38 <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->

bijington commented:
Sample incoming at some point. I can't seem to build at the moment though, I suspect I need a VS update

pictos commented on src/CommunityToolkit.Maui.Markup/VisualElementExtensions.cs:
```diff
@@ -3,114 +3,169 @@
+	{
+		foreach (var behavior in behaviors)
+		{
+			element.Behaviors.Add(behavior);
```
`Behaviors` is of type IList

bijington commented:
Added a very basic settings page to expose this new `Behaviors` method in a sample. Also removed the AbsoluteLayout page as it wasn't actually accessible in the app and move some of it's logic out to the settings page.

@brminnick let me know your thoughts. I did have to pull a dependency on the toolkit in order to consume a behavior

brminnick reviewed (changes requested):
Thanks Shaun!! 

I think it's a great idea to use the `CommunityToolkit.Maui` NuGet package in the sample app! In the `CommunityToolit.Maui` sample app, we use `CommunityToolit.Maui.Markup`, and it makes a lot of sense to reciprocate that here.

## Updates

Just FYI - I tweaked the sample app a bit:
- Implement `CommunityToolkit.Mvvm`
  - `BaseViewModel` now inherits from `CommunityToolkit.Mvvm.ComponentModel.ObservableObject`
  - `IsListRefreshing` is now auto-implemented using `[ObservableProperty]`
- Implement Maui.Essentials Interfaces
- Replace `MainThread` with `IDispatcher`
  - `MainThread` is deprecated in the next release of .NET MAUI
- Add `InvalidStyle` to `NumericValidationBehavior`
- Add `ValidStyle` to `NumericValidationBehavior` 
- Remove `ISettingsService`
  - Matches the precedence set by `HackerNewsApiService` which doesn't use an interface
- `SettingsService`
  - Use `IPreferences` to save the user's preference for `NumberOfTopStoriesToFetch`
  - Use `Math.Clamp` to set the upper + lower bounds of `NumberOfTopStoriesToFetch`

## Request 

I only have two requests, both in the Sample app:
- Tweak `SettingsPage` a bit to look good on all platforms
- In `SettingsPage`, add a `Label` below the `Entry` to let users now the minimum + maximum bounds for `NumberOfTopStoriesToFetch` 

Here's how it currently looks on macOS, iOS and Android:

| macOS | iOS | Android |
| ------- | ----| --------| 
| <img width="1042" alt="macOS" src="https://user-images.githubusercontent.com/13558917/160299995-4de4d634-0c0e-4848-8f08-6485e70be7ad.png"> | ![iOS](https://user-images.githubusercontent.com/13558917/160300002-61bc6d87-99fb-4ece-94ff-b79f7f39342f.png) | ![Android](https://user-images.githubusercontent.com/13558917/160300028-26ce0089-1753-4fe4-9def-4ef96602415d.png) |

bijington commented:
You can clearly see the environment I tested this on 😄. I am starting to regret using the AbsoluteLayout for this page, I expected laying it out could be painful but I wanted to justify removing the samples page.

Perhaps I should switch to using a Grid and expose some of its related extensions instead or something we don't have a sample for and save AbsoluteLayout for some it is really suited to (I don't know what yet)

Happy to proceed and fix this stuff. Thanks for all the updates/fixes

brminnick commented:
If you're able to devote the time, I'd love to continue using `AbsoluteLayout` for `SettingsPage` 💯 

We're already demonstrating the use of `Grid` with C# Markup Extensions in `StoryDataTemplate`.

I've found that `AbsoluteLayout` is a little confusing to learn at first, but it is super powerful once you wrap your mind around it.

bijington commented:
Yep I'll happily take a look. It'll help me learn it better 😄

brminnick commented on samples/CommunityToolkit.Maui.Markup.Sample/Pages/NewsPage.cs:
```diff
@@ -46,6 +66,8 @@ protected override void OnAppearing()
 			refreshView.IsRefreshing = true;
 		}
 
+		BindingContext.RefreshCommand.Execute(null);
```
Strange - yea, `refreshView.IsRefreshing = true` does the same thing as `BindingContext.RefreshCommand.Execute(null);`.

We should remove it 💯 

```suggestion
```

brminnick commented on samples/CommunityToolkit.Maui.Markup.Sample/Pages/SettingsPage.cs:
```diff
@@ -0,0 +1,55 @@
+					})
+					.Bind(Entry.TextProperty, nameof(SettingsViewModel.NumberOfTopStoriesToFetch)),
+
+				new Label
+				{
+					Text = $"Top Stories To Fetch", TextColor = ColorConstants.PrimaryTextColor, HorizontalTextAlignment = TextAlignment.Center,
+					FontAttributes = FontAttributes.Italic, FontSize = 12
+				}
+					.Bind<Label, int, int, string>(
+						Label.TextProperty,
+						binding1: new Binding { Source = SettingsService.MinimumStoriesToFetch },
+						binding2: new Binding { Source = SettingsService.MaximumStoriesToFetch },
+						convert: ((int minimum, int maximum) values) => string.Format(CultureInfo.CurrentUICulture, "The number must be between {0} and {1}.", values.minimum, values.maximum))
+					.LayoutFlags(AbsoluteLayoutFlags.XProportional | AbsoluteLayoutFlags.WidthProportional)
+					.LayoutBounds(0,90,1,40),
```
Technically the multi-binding isn't needed since it is a `OneTime` binding (ie the binding source is constant).

However, I like keeping the multi binding to demonstrate how to use multi binding it in the sample app!

(Here's what it would look like without the binding)
```cs
				new Label { Text = $"The number must be between SettingsService.MinimumStoriesToFetch and SettingsService.MaximumStoriesToFetch.", TextColor = ColorConstants.PrimaryTextColor }
					.LayoutFlags(AbsoluteLayoutFlags.XProportional | AbsoluteLayoutFlags.WidthProportional)
					.LayoutBounds(0,90,1,40)
					.TextCenter()
					.Font(size: 12, italic: true)
```

bijington replied:
I've got so much to learn! Thanks for all these suggestions.

Yes I like the idea of showing off a multi binding here despite the value not changing. I did also consider whether we should localise eventually to show that concept off too. Not in scope of this PR though.

brminnick reviewed (commented):
Thanks Shaun! It's looking good!!

I found a couple more places where we can leverage C# Markup Extension Methods!

bijington commented on samples/CommunityToolkit.Maui.Markup.Sample/Pages/SettingsPage.cs:
```diff
@@ -0,0 +1,55 @@
+			{
+				new Label { Text = "Top Stories To Fetch", TextColor = ColorConstants.PrimaryTextColor, HorizontalTextAlignment = TextAlignment.Center }
+					.LayoutFlags(AbsoluteLayoutFlags.XProportional | AbsoluteLayoutFlags.WidthProportional)
+					.LayoutBounds(0,0,1,40),
```
What's the difference between `CenterHorizontal` and `TextCenter`?

bijington replied:
Oh `LayoutOptions` vs `TextAlignment` thanks

brminnick commented on samples/CommunityToolkit.Maui.Markup.Sample/Pages/SettingsPage.cs:
```diff
@@ -0,0 +1,54 @@
+					})
+					.Bind(Entry.TextProperty, nameof(SettingsViewModel.NumberOfTopStoriesToFetch)),
+
+				new Label { TextColor = ColorConstants.PrimaryTextColor, HorizontalTextAlignment = TextAlignment.Center }
```
We can remove `HorizontalTextAlignment = TextAlignment.Center` now that we've added `.TextCenter()` below
```suggestion
				new Label { TextColor = ColorConstants.PrimaryTextColor }

```

brminnick commented on samples/CommunityToolkit.Maui.Markup.Sample/Pages/SettingsPage.cs:
```diff
@@ -0,0 +1,54 @@
+		{
+			Children =
+			{
+				new Label { Text = "Top Stories To Fetch", TextColor = ColorConstants.PrimaryTextColor, HorizontalTextAlignment = TextAlignment.Center }
+					.LayoutFlags(AbsoluteLayoutFlags.XProportional | AbsoluteLayoutFlags.WidthProportional)
+					.LayoutBounds(0,0,1,40),
```
This can use `.TextCenter()` to replace `HorizontalTextAlignment = TextAlignment.Center`

```suggestion
				new Label { Text = "Top Stories To Fetch", TextColor = ColorConstants.PrimaryTextColor }
					.LayoutFlags(AbsoluteLayoutFlags.XProportional | AbsoluteLayoutFlags.WidthProportional)
					.LayoutBounds(0,0,1,40)
					.TextCenter(),

```

bijington replied:
Sorry I should have checked it also

bijington commented:
Oh of course this requires docs!!

brminnick reviewed (approved):
Thanks Shaun!!
